### PR TITLE
fix: allow the purge method to purge cache

### LINKED ISSUE #6766: bug: The HTTP method must be selected as ALL to use the PURGE function during the use of the proxy cache.
### Current Behavior

When I want to purge to cache I got an "HTTP 404 Not Found" error. 
When I looked a little bit, I saw that the purge method did not work when I only allowed the GET and HEAD methods as the HTTP method. I agree, this is normal behavior. 
But the PURGE method is not supported by API or interface. I realized that I should choose ALL as the HTTP method instead. 
In short, the HTTP method must be selected as ALL to use the PURGE function during the use of the proxy cache.

Ps.sorry about that I'm not sure if this is a bug or a new feature maybe both.

### Expected Behavior

The PURGE method should be supported within the HTTP method options.

### Error Logs

```
$ curl http://10.61.11.4:9080/hello?cache=2222 -H "Host: test.example.com" -X PURGE -v
*   Trying 10.61.11.4:9080...
* TCP_NODELAY set
* Connected to 10.61.11.4 (10.61.11.4) port 9080 (#0)
> PURGE /hello?cache=2222 HTTP/1.1
> Host: test.example.com
> User-Agent: curl/7.68.0
> Accept: */*
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 Not Found
< Date: Thu, 31 Mar 2022 14:27:06 GMT
< Content-Type: text/plain; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
<
{"error_msg":"404 Route Not Found"}
* Connection #0 to host 10.61.11.4 left intact

### Steps to Reproduce

The error reproduces when the PURGE method is called in a configuration that does not have ALL selected as the HTTP method.

### Environment

- APISIX version (run `apisix version`): 2.11.0
- Operating system (run `uname -a`): ` Linux mt-tap6-linlb03 5.4.0-97-generic #110~18.04.1-Ubuntu SMP Mon Jan 17 20:50:31 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux `
- OpenResty / Nginx version (run `openresty -V` or `nginx -V`): `nginx version: openresty/1.19.9.1
built with OpenSSL 1.1.1k  25 Mar 2021 (running with OpenSSL 1.1.1l  24 Aug 2021)
TLS SNI support enabled
configure arguments: --prefix=/usr/local/openresty/nginx --with-cc-opt='-O2 -DNGX_LUA_ABORT_AT_PANIC -I/usr/local/openresty/zlib/include -I/usr/local/openresty/pcre/include -I/usr/local/openresty/openssl111/include' --add-module=../ngx_devel_kit-0.3.1 --add-module=../echo-nginx-module-0.62 --add-module=../xss-nginx-module-0.06 --add-module=../ngx_coolkit-0.2 --add-module=../set-misc-nginx-module-0.32 --add-module=../form-input-nginx-module-0.12 --add-module=../encrypted-session-nginx-module-0.08 --add-module=../srcache-nginx-module-0.32 --add-module=../ngx_lua-0.10.20 --add-module=../ngx_lua_upstream-0.07 --add-module=../headers-more-nginx-module-0.33 --add-module=../array-var-nginx-module-0.05 --add-module=../memc-nginx-module-0.19 --add-module=../redis2-nginx-module-0.15 --add-module=../redis-nginx-module-0.3.7 --add-module=../ngx_stream_lua-0.0.10 --with-ld-opt='-Wl,-rpath,/usr/local/openresty/luajit/lib -L/usr/local/openresty/zlib/lib -L/usr/local/openresty/pcre/lib -L/usr/local/openresty/openssl111/lib -Wl,-rpath,/usr/local/openresty/zlib/lib:/usr/local/openresty/pcre/lib:/usr/local/openresty/openssl111/lib' --with-pcre-jit --with-stream --with-stream_ssl_module --with-stream_ssl_preread_module --with-http_v2_module --without-mail_pop3_module --without-mail_imap_module --without-mail_smtp_module --with-http_stub_status_module --with-http_realip_module --with-http_addition_module --with-http_auth_request_module --with-http_secure_link_module --with-http_random_index_module --with-http_gzip_static_module --with-http_sub_module --with-http_dav_module --with-http_flv_module --with-http_mp4_module --with-http_gunzip_module --with-threads --with-stream --with-http_ssl_module`
- etcd version, if relevant (run `curl http://127.0.0.1:9090/v1/server_info`): `{"id":"aa24eed2-fb04-4e43-ad54-daa26c44e429","version":"2.11.0","etcd_version":"3.4.0","up_time":18802,"last_report_time":1648737880,"hostname":"mt-tap6-linlb03","boot_time":1648719100}`
- APISIX Dashboard version, if relevant:
- Plugin runner version, for issues related to plugin runners:
- LuaRocks version, for installation issues (run `luarocks --version`): `/usr/local/bin/luarocks 3.8.0`

### PR DESCRIPTION
**Pull request by soulbird**
### Description

Allows adding purge method restrictions to routes. 
The purge method is used in the proxy-cache plugin and is usually used to flush the cache.

Fixes #6766

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [ ] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by membphis — COMMENTED** [review]
need to add a new test case:

1. set route with `PURGE` method
2. hit the route by http request (PURGE method)
3. check the response context

**Comment by soulbird** [comment]
> need to add a new test case:
> 
> 1. set route with `PURGE` method
> 2. hit the route by http request (PURGE method)
> 3. check the response context

Route matching has added test case here https://github.com/api7/lua-resty-radixtree/blob/master/t/sanity.t#L347

**Review by spacewander — DISMISSED** [review]

**Comment by bzp2010** [comment]
https://github.com/apache/apisix/blob/97bcb565075e59372842973844216c129af5d812/apisix/plugins/proxy-rewrite.lua#L27-L34

https://github.com/apache/apisix/blob/97bcb565075e59372842973844216c129af5d812/apisix/plugins/proxy-rewrite.lua#L50-L54

The `proxy-rewrite` plugin support method rewrite also, should we modify it?

**Comment by soulbird** [comment]
> https://github.com/apache/apisix/blob/97bcb565075e59372842973844216c129af5d812/apisix/plugins/proxy-rewrite.lua#L27-L34
> 
> https://github.com/apache/apisix/blob/97bcb565075e59372842973844216c129af5d812/apisix/plugins/proxy-rewrite.lua#L50-L54
> 
> The `proxy-rewrite` plugin support method rewrite also, should we modify it?

We support the `PURGE` method for the proxy-cache plugin. But `PURGE` is not a standard method, we don't need to support proxy `PURGE` method

**Comment by membphis** [comment]
> need to add a new test case:
> 
> 1. set route with `PURGE` method
> 2. hit the route by http request (PURGE method)
> 3. check the response context

@soulbird What is your opinion?

**Comment by membphis** [comment]
> https://github.com/api7/lua-resty-radixtree/blob/master/t/sanity.t#L347

You need to use real test cases, we may find bug in real test case.

**Review by membphis — CHANGES_REQUESTED** [review]
missing test case(PURGE method)

**Comment by soulbird** [comment]
> > need to add a new test case:
> > 
> > 1. set route with `PURGE` method
> > 2. hit the route by http request (PURGE method)
> > 3. check the response context
> 
> Route matching has added test case here https://github.com/api7/lua-resty-radixtree/blob/master/t/sanity.t#L347



> > https://github.com/api7/lua-resty-radixtree/blob/master/t/sanity.t#L347
> 
> You need to use real test cases, we may find bug in real test case.



> > https://github.com/api7/lua-resty-radixtree/blob/master/t/sanity.t#L347
> 
> You need to use real test cases, we may find bug in real test case.

OK, I will add test cases related route matching. And the test cases related resppnse context  has already added in proxy-cache plugin

File: t/plugin/proxy-cache/disk.t [hunk 1]
```diff
@@ -706,3 +706,115 @@ GET /t
 --- error_code: 400
 --- response_body eval
 qr/failed to check the configuration of plugin proxy-cache err/
+
+
+
+=== TEST 28: check purge method (route no purge method)
```
Comment (membphis) [inline t/plugin/proxy-cache/disk.t:712]: It is not a good idea to add new test cases to the `proxy-cache` plugin.

We should use a new test file.

**Review by spacewander — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Pull request merged**